File: 2019/day04/day04.hs
#!/usr/bin/env stack
-- stack --resolver lts-14.16 script
module Day04 where

import System.Environment (getArgs)
import Data.List.Split (splitOn)
import Data.List (group)

monotonic :: (Ord a) => [a] -> Bool
monotonic xs = all (uncurry (<=)) $ zip xs (tail xs)

hasConsecutive :: (Eq a) => [a] -> Bool
hasConsecutive = any ((> 1) . length) . group

hasAdjacent :: (Eq a) => [a] -> Bool
hasAdjacent = any ((== 2) . length) . group

star1 :: (Ord a, Eq a) => [[a]] -> Int
star1 = length . filter hasConsecutive . filter monotonic

star2 :: (Ord a, Eq a) => [[a]] -> Int
star2 = length . filter hasAdjacent . filter monotonic

main :: IO ()
main = do
  args <- getArgs
  let
    filename = case args of
      (x:_) -> x
      _     -> "input.txt"
  [start, end] <- map (read :: String -> Int) . splitOn "-" <$> readFile filename
  let input = map show [start..end]
  print $ star1 input
  print $ star2 input

File: 2019/day04/day04.rs
#!/usr/bin/env run-cargo-script
use std::collections::HashMap;

fn monotonic(s: &str) -> bool {
    for (i, j) in s.chars().zip(s.chars().skip(1)) {
        if i > j {
            return false;
        }
    }
    true
}

// Yes, making a hashmap is overkill for this problem
fn freqs(s: &str) -> HashMap<char, i32> {
    let mut res = HashMap::new();
    for i in s.chars() {
        res.entry(i).and_modify(|e| *e += 1).or_insert(1);
    }
    res
}

fn hasConsecutive(s: &str) -> bool {
    freqs(s).iter().any(|(_, i)| *i >= 2)
}

fn hasAdjacent(s: &str) -> bool {
    freqs(s).iter().any(|(_, i)| *i == 2)
}

fn star1(start: i32, end: i32) -> usize {
    (start..=end)
        .map(|i| format!("{}", i))
        .filter(|i| monotonic(i))
        .filter(|i| hasConsecutive(i))
        .count()
}

fn star2(start: i32, end: i32) -> usize {
    (start..=end)
        .map(|i| format!("{}", i))
        .filter(|i| monotonic(i))
        .filter(|i| hasAdjacent(i))
        .count()
}

fn main() {
    let args: Vec<_> = std::env::args().collect();
    let filename = args.get(1).map(|s| &s[..]).unwrap_or("input.txt");
    let file_contents = std::fs::read_to_string(filename).unwrap();
    let mut input = file_contents.trim().split("-");
    let start = input.next().unwrap().parse().unwrap();
    let end = input.next().unwrap().parse().unwrap();

    println!("{}", star1(start, end));
    println!("{}", star2(start, end));
}
